test/e2e_env/multizone/reachablebackends/reachablebackends_meshservices.go

```diff
@@ -0,0 +1,179 @@
+			)).
+			Install(YamlK8s(fmt.Sprintf(`
+apiVersion: kuma.io/v1alpha1
+kind: HostnameGenerator
+metadata:
+  name: local-k8s-mesh-services
+  namespace: %s
+  labels:
+    kuma.io/origin: zone
+spec:
+  template: '{{ .DisplayName }}.{{ .Namespace }}.svc.{{ .Zone }}.mesh.local'
+  selector:
+    meshService:
+      matchLabels:
+        kuma.io/origin: zone
+        kuma.io/env: kubernetes
+`, Config.KumaNamespace))).
+			Setup(multizone.KubeZone1)
+		Expect(err).ToNot(HaveOccurred())
+
+		err = NewClusterSetup().
+			Install(NamespaceWithSidecarInjection(namespace)).
+			Install(testserver.Install(
+				testserver.WithName("other-zone-test-server"),
+				testserver.WithNamespace(namespace),
+				testserver.WithMesh(meshName),
+				testserver.WithEchoArgs("echo", "--instance", "other-zone-test-server"),
+			)).
+			Setup(multizone.KubeZone2)
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	AfterEachFailure(func() {
+		DebugUniversal(multizone.Global, meshName)
+		DebugKube(multizone.KubeZone1, meshName, namespace)
+		DebugKube(multizone.KubeZone2, meshName, namespace)
+	})
+
+	E2EAfterAll(func() {
+		Expect(multizone.KubeZone1.TriggerDeleteNamespace(namespace)).To(Succeed())
+		Expect(multizone.KubeZone2.TriggerDeleteNamespace(namespace)).To(Succeed())
+		Expect(multizone.Global.DeleteMesh(meshName)).To(Succeed())
+	})
+
+	It("should be able to connect to k8s VIP, using legacy cluster, if reachable backends isn't set", func() {
+		Eventually(func(g Gomega) {
+			// when
+			resp, err := client.CollectEchoResponse(
+				multizone.KubeZone1, "client-without-reachable", "local-test-server.mesh-service-reachable-backends.svc.kuma-1.mesh.local",
+				client.FromKubernetesPod(namespace, "client-without-reachable"),
+			)
+			// then
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(resp.Instance).To(HavePrefix("local-test-server"))
+		}, "30s", "500ms", MustPassRepeatedly(10)).Should(Succeed())
+
+		Eventually(func(g Gomega) {
+			pod, err := PodNameOfApp(multizone.KubeZone1, "client-without-reachable", namespace)
+			g.Expect(err).ToNot(HaveOccurred())
+			stdout, err := multizone.KubeZone1.GetKumactlOptions().RunKumactlAndGetOutput("inspect", "dataplane", pod+"."+namespace, "--type=clusters", fmt.Sprintf("--mesh=%s", meshName))
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(stdout).To(ContainSubstring(fmt.Sprintf("local-test-server_%s_svc_80", namespace)))
+			g.Expect(stdout).ToNot(ContainSubstring("_msvc_"))
+		}, "10s", "1s").Should(Succeed())
+	})
+
+	It("should not be able to connect to cross-zone MeshService if reachable backends isn't set", func() {
+		Consistently(func(g Gomega) {
+			// when
+			resp, err := client.CollectFailure(
+				multizone.KubeZone1, "client-without-reachable", "other-zone-test-server.mesh-service-reachable-backends.svc.kuma-2.mesh.local",
+				client.FromKubernetesPod(namespace, "client-without-reachable"),
+			)
+			// then
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(resp.Exitcode).To(Equal(6))
```

Comment:
A comment explaning the exit code is helpful: "exit code 6: Could Not Resolve Host"

Reply:
Probably we could replace these everywhere with a `framework.CurlExitCodes` type